Role Strategy Plugin GSoC Phase 1 blog (#2335)
jenkins-infra/jenkins.io: content/blog/2019/06/2019-06-14-performance-testing-jenkins.adoc

---
layout: post
title: "Performance Testing Framework for Jenkins Plugins"
tags:
- jenkins
- jmh
- plugins
- benchmark
- performance
- gsoc
author: abhyudayasharma
---

= Performance Testing Framework for Jenkins Plugins

> 

Reviewer: Add short self and project introduction?

Author: ![image](https://user-images.githubusercontent.com/11471599/59514604-d50e2700-8eda-11e9-82da-9fdbbfd67781.png)

---
layout: post

> title: "Performance Testing Framework for Jenkins Plugins"

Reviewer: Or Micro-benchmarking, WDYT?

Author: Changed it to 'Microbenchmarking framework for Jenkins`

tags:

> - jenkins

Reviewer: Not needed. Everything is about Jenkins in this blog

- jmh
- plugins
- benchmark

> - performance

Reviewer: ```suggestion
- performance
- developer
```

> - gsoc

Reviewer: ```suggestion
- gsoc
- gsoc2019
```

author: abhyudayasharma
---

= Performance Testing Framework for Jenkins Plugins

My work for the first phase of my Google Summer of Code project to improve the performance of
the Role Strategy Plugin involved creating a framework for running benchmarks in Jenkins Plugins.

> We chose link:http://openjdk.java.net/jeps/230[Java Microbenchmark Harness] for running the benchamrks.

Reviewer: Would be nice to expand it a bit to highlight the community value. Then the technology side could go to the second paragraph along with the reference to releases which include the framework

The framework works by starting a temporary Jenkins instance for each fork of the JMH benchmark, just
like `JenkinsRule` from Jenkins Test Harness.

== Creating Benchmarks

Benchmarks are run directly from your JUnit Tests which allows you to fail builds on the fly and easily run
benchmarks from your IDE, just like unit tests. You can easily configure your benchmarks by either using your

> Java methods, or by using Jenkins Configuration-as-Code plugin and passing the path to your YAML file.

Reviewer: An example would be great (maybe just as a link)

Author: I've added a section for Configuration as Code, is that enough or add should I add a link to a benchmark using CasC in Role Strategy.

=== Writing benchmarks

A reference to the Jenkins instance is available through either `JmhBenchmarkState#getJenkins()` or through
`Jenkins.getInstance()` like you would otherwise do. `JmhBenchmarkState` provides `setup()` and `tearDown` methods
which can be overridden to configure the Jenkins instance according to your benchmark's requirements.

=== Running the benchmarks

The benchmarks can be run through JUnit tests. From a test method, you can use the `OptionsBuilder` provided by JMH to
configure your benchmarks. Benchmark reports generated (in JSON) can be visualized using the

> link:https://wiki.jenkins.io/display/JENKINS/JMH+Report+Plugin[JMH Report Plugin].

Reviewer: ```suggestion
plugin:jmh-report[JMH Report Plugin].
```

Will refer to https://plugins.jenkins.io/jmh-report

like a JUnit test:

[source, java]
----
public class BenchmarkRunner {
    @Test
    public void runJmhBenchmarks() throws Exception {
        ChainedOptionsBuilder options = new OptionsBuilder()
                .mode(Mode.AverageTime)
                .forks(2)
                .result("jmh-report.json");

        // Automatically detect benchmark classes annotated with @JmhBenchmark
        BenchmarkFinder.findBenchmarks(options);
        new Runner(options.build()).run();
    }
}
----

This allows us to automatically find benchmarks and run them through a JUnit test.
Since `BenchmarkRunner` class name does not qualify as a test according to Maven surefire plugin's
naming conventions, the benchmarks will not interfere with your JUnit tests.
Now, you can write your benchmarks:

**Without any special setup**:

[source,java]
----
@JmhBenchmark
public class JmhStateBenchmark {
    public static class MyState extends JmhBenchmarkState {
    }

    @Benchmark
    public void benchmark(MyState state) {
        // benchmark code goes here
        state.getJenkins().setSystemMessage("Hello world");
    }
}
----

> // TODO: Add docs for configuration as code after configuration-as-code#921 is merged

Reviewer: Maybe all configuration-as-code logic could go to a separate section. It is an interesting content, and it worth highligting it. You can start drafting the doc IMHO. Even if the delivery gets delayed, in the worst case we could just move it to a separate blogpost.

As a part of this project, benchmarks have been created in the Role Strategy Plugin which show
configuring the instances and the benchmarks for various situations. You may find them 
link:https://github.com/jenkinsci/role-strategy-plugin/tree/master/src/test/java/jmh/benchmarks[here].

== Integrating with your plugin

This framework has now been integrated into Jenkins Test Harness and is available for everyone 
to use by just upgrading the version of the Plugin-POM. To use Configuration-as-Code to set up your Jenkins
instances that will be started for the benchmarks, you also need to add a Maven dependency to the 
`configuration-as-code` plugin.

Plugins built on ci.jenkins.io can also use the `runBenchmarks()` method after the `buildPlugin()` step in your
Jenkinsfile which is now available in the Jenkins Pipeline library. This function also accepts the path to your
generated JMH benchmark reports as an optional parameter and archives the benchmark results. The benchmark
builds are currently throttled because of the limited availability of `highmem` nodes. Running benchmarks in 
pull request builds allows you to constantly monitor the performance implications of a given change.

> 

Reviewer: Add Jenkinsfile Sample from Role Strategy?

---
layout: post
title: "Microbenchmarking Framework for Jenkins Plugins"
tags:
- jmh
- plugins
- benchmark
- performance
- developer
- gsoc
- gsoc2019
author: abhyudayasharma
---

= Performance Testing Framework for Jenkins Plugins

> I have been working on improving the performance of the Role Strategy Plugin as a part of my Google Summer of

Reviewer: Add a link to your project?

Code project. Since there was no existing way to measure performance and do benchmarks on Jenkins Plugins,
my work for the first phase of this Google Summer of Code project was to create a framework for running
benchmarks in Jenkins Plugins with a Jenkins instance available. To make our job a bit easier,
we chose link:http://openjdk.java.net/jeps/230[Java Microbenchmark Harness] for running these benchmarks. This
allows us to reliably measure performance of our time-critical functions and will help make Jenkins perform faster

> for everyone. The blogpost below shows how to run a benchmark in your plugin.

Reviewer: ```suggestion
for everyone. 

The micro-benchmarking framework has been released recently as a part of [Jenkins Unit Test Harness](https://github.com/jenkinsci/jenkins-test-harness) 2.50. 
The blogpost below shows how to run a benchmark in your plugin.
```

> == Creating Benchmarks

Reviewer: ```suggestion
== Summary
```

The framework runs works by starting a temporary Jenkins instance for each fork of the JMH benchmark,
just like `JenkinsRule` from Jenkins Test Harness. Benchmarks are run directly from your JUnit Tests which allows
you to fail builds on the fly and easily run benchmarks from your IDE, just like unit tests. You can easily
configure your benchmarks by either using your Java methods, or by using Jenkins Configuration-as-Code plugin
and passing the path to your YAML file.

To run benchmarks from your plugins, you either need to do the following:

* bump up the minimum required Jenkins version to 2.60.3 or above
* bump Plugin-POM to a version >= 3.46 or manually upgrade to Jenkins Test Harness >= 2.51.

=== Writing benchmarks

> A reference to the Jenkins instance is available through either `JmhBenchmarkState#getJenkins()` or through

Reviewer: Implementation detail instead of guidelines. Move code from `Sample benchmarks` here?

`Jenkins.getInstance()` like you would otherwise do. `JmhBenchmarkState` provides `setup()` and `tearDown()` methods
which can be overridden to configure the Jenkins instance according to your benchmark's requirements.

=== Running the benchmarks

The benchmarks can be run through JUnit tests. From a test method, you can use the `OptionsBuilder` provided by JMH to
configure your benchmarks. Benchmark reports generated (in JSON) can be visualized using the

> plugin:jmh-report[JMH Report Plugin].

Reviewer: Add a section for `viewing benchmarks`. Reference the this plugin or http://jmh.morethan.io/ (reference diff options there?). Maybe it makes sense to add a screenshot for Role Strategy

> ==== Automatic detection of benchmark classes

Reviewer: Implementation detail. Maybe group all of them to a special "Tips and Tricks" section?

Classes containing benchmarks are found automatically by the `BenchmarkFinder`
when annotated with `@JmhBenchmark`.

NOTE: Benchmark methods need to be annotated by `@Benchmark` for JMH to detect them.

==== Running benchmarks from Maven

To easily run benchmrks from Maven, a Maven profile to run the benchmarks has been created
and is available starting Plugin-POM version 3.45. You can then run your benchmarks from the
command line using `mvn test -Dbenchmark`. To make it easier for plugins hosted on ci.jenkins.io,

> a build step has been added in Jenkins pipeline library which automatically runs benchmarks

Reviewer: Add a "Running benchmarks in ci.jenkins.io" section here?

like a JUnit test:

[source, java]
----
public class BenchmarkRunner {
    @Test
    public void runJmhBenchmarks() throws Exception {
        ChainedOptionsBuilder options = new OptionsBuilder()
                .mode(Mode.AverageTime)
                .forks(2)
                .result("jmh-report.json");

        // Automatically detect benchmark classes annotated with @JmhBenchmark
        new BenchmarkFinder(getClass()).findBenchmarks(options);
        new Runner(options.build()).run();
    }
}
----

This allows us to automatically find benchmarks and run them through a JUnit test.
Since `BenchmarkRunner` class name does not qualify as a test according to Maven surefire plugin's
naming conventions, the benchmarks will not interfere with your JUnit tests.
Now, you can write your benchmarks:

**Without any special setup**:

[source,java]
----
@JmhBenchmark
public class JmhStateBenchmark {
    public static class MyState extends JmhBenchmarkState {
    }

    @Benchmark
    public void benchmark(MyState state) {
        // benchmark code goes here
        state.getJenkins().setSystemMessage("Hello world");
    }
}
----

> 

Reviewer: Add sample output (screenshot/console text)?

**More Samples**:

As a part of this project, benchmarks have been created in the Role Strategy Plugin which show
configuring the instances and the benchmarks for various situations. You may find them 
link:https://github.com/jenkinsci/role-strategy-plugin/tree/master/src/test/java/jmh/benchmarks[here].

> == Using Configuration as Code

Reviewer: ```suggestion
== Using Configuration as Code in benchmarks
```

----
<dependency>
    <groupId>io.jenkins</groupId>
    <artifactId>configuration-as-code</artifactId>
    <version>1.21</version>
    <optional>true</optional>
</dependency>
<dependency>
    <groupId>io.jenkins</groupId>
    <artifactId>configuration-as-code</artifactId>
    <version>1.21</version>
    <classifier>tests</classifier>
    <scope>test</scope>
</dependency>
----

// TODO: Verify version after configuration-as-code#921 is released

Now configuring a benchmark is as simple as providing path to your YAML file and specifying the class
containing the benchmark state.

[source,java]
----
@JmhBenchmark
public class SampleBenchmark {
    public static class MyState extends CascJmhBenchmarkState {
        @Nonnull
        @Override
        protected String getResourcePath() {
            return "config.yml";
        }

        @Nonnull
        @Override
        protected Class<?> getEnclosingClass() {
            return SampleBenchmark.class;
        }
    }

    @Benchmark
    public void benchmark(MyState state) {
        Jenkins jenkins = state.getJenkins(); // jenkins is configured and ready to be benchmarked.
        // your benchmark code goes here...
    }
}
----

> == Integrating with your plugin

Reviewer: Looks like all the text will go to other sections

Reviewer: ```suggestion
== Running benchmarks
```

The first paragraph below is already written above, could be removed IMHO

This framework has now been integrated into Jenkins Test Harness and is available for everyone 
to use by just upgrading the version of the Plugin-POM. To use Configuration-as-Code to set up your Jenkins
instances that will be started for the benchmarks, you also need to add a Maven dependency to the 
`configuration-as-code` plugin.

Plugins built on ci.jenkins.io can also use the `runBenchmarks()` method after the `buildPlugin()` step in your
Jenkinsfile which is now available in the Jenkins Pipeline library. This function also accepts the path to your
generated JMH benchmark reports as an optional parameter and archives the benchmark results. The benchmark
builds are currently throttled because of the limited availability of `highmem` nodes. Running benchmarks in 
pull request builds allows you to constantly monitor the performance implications of a given change. For example,
the Jenkinsfile from Role Strategy Plugin:
[source, groovy]
----
buildPlugin()
runBenchmarks('jmh-report.json')
----

> == Demos and Presentation Slides

Reviewer 2: Do you really need this section at this stage? (Maybe after the first demo)

++++
<div style="text-align: center; margin: 30px 0px;">
<iframe src="https://docs.google.com/presentation/d/e/2PACX-1vQXca_ZQNwI4vQ25Nw7lMSnSh4WBwKbC9VltT-7tOjS8cE69zMb2bgEbhgwurb1xA/embed?start=false&loop=false&delayms=5000" frameborder="0" width="640" height="389" allowfullscreen="true" mozallowfullscreen="true" webkitallowfullscreen="true" style="margin-bottom: 20px;"></iframe>
<iframe width="640" height="389" src="https://www.youtube.com/embed/sr28UADG1AE" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
</div>

> ++++

Reviewer: ```suggestion
++++

== Links and feedback

If you have any questions about the framework, feel free to reach out to me in the [Gitter chat](LINK) or the developer mailing list (thread about plugin benchmarking). You can also join the Jenkins Online Meetup on June 25 where I will be presenting the framework.

Other links ...

* Video
* Project page
...

```

Reviewer: @oleg-nenashev to supply a Jenkins Online Meetup link

---
layout: post
title: "Micro-benchmarking Framework for Jenkins Plugins"
tags:
- jmh
- plugins
- benchmark
- performance
- developer
- gsoc
- gsoc2019
author: abhyudayasharma
---

= Micro-benchmarking Framework for Jenkins Plugins

> I have been working on improving the performance of the Role Strategy Plugin as a part of my link:https://summerofcode.withgoogle.com/projects/#6612887823450112[Google Summer of Code project].

Reviewer: Fine with me, but maybe https://jenkins.io/projects/gsoc/2019/role-strategy-performance/ is a better link

Reviewer: In such case `link:/projects/gsoc/2019/role-strategy-performance/[Google Summer of Code project]` could be used

Since there was no existing way to measure performance and do benchmarks on Jenkins Plugins,

> my work for the first phase of this Google Summer of Code project was to create a framework for running

Reviewer: ```suggestion
my work for the first phase of this project was to create a framework for running
```

benchmarks in Jenkins Plugins with a Jenkins instance available. To make our job a bit easier,
we chose link:http://openjdk.java.net/jeps/230[Java Microbenchmark Harness] for running these benchmarks. This
allows us to reliably measure performance of our time-critical functions and will help make Jenkins perform faster
for everyone.

> The micro-benchmarking framework was recently released in the Jenkins Unit Test Harness.

Reviewer: ```suggestion
The micro-benchmarking framework was recently released in the Jenkins Unit Test Harness 2.50.
```

Maybe also makes sense to link https://github.com/jenkinsci/jenkins-test-harness/releases/tag/jenkins-test-harness-2.50

                .result("jmh-report.json");

        // Automatically detect benchmark classes annotated with @JmhBenchmark
        new BenchmarkFinder(getClass()).findBenchmarks(options);
        new Runner(options.build()).run();
    }
}
----

== Sample benchmarks

Now, you can write your first benchmark:

**Without any special setup**

[source,java]
----
@JmhBenchmark
public class JmhStateBenchmark {
    public static class MyState extends JmhBenchmarkState {
    }

    @Benchmark
    public void benchmark(MyState state) {
        // benchmark code goes here
        state.getJenkins().setSystemMessage("Hello world");
    }
}
----

**Using Configuration as Code**

To use configuration as code, apart from the dependencies above you also need to add the following
to your `pom.xml`:

[source,xml]
----
<dependency>
    <groupId>io.jenkins</groupId>
    <artifactId>configuration-as-code</artifactId>
    <version>1.21</version>
    <optional>true</optional>
</dependency>
<dependency>
    <groupId>io.jenkins</groupId>
    <artifactId>configuration-as-code</artifactId>
    <version>1.21</version>
    <classifier>tests</classifier>
    <scope>test</scope>
</dependency>
----

> // TODO: Verify version after configuration-as-code#921 is released

Reviewer: done?

        // your benchmark code goes here...
    }
}
----

**More Samples**

As a part of this project, a few benchmarks have been created in the Role Strategy Plugin which show
configuring the instances for various situations. You can find them 
link:https://github.com/jenkinsci/role-strategy-plugin/tree/master/src/test/java/jmh/benchmarks[here].

== Integrating with your plugin

This framework has now been integrated into Jenkins Test Harness and is available for everyone 
to use by just upgrading the version of the Plugin-POM. To use Configuration-as-Code to set up your Jenkins
instances that will be started for the benchmarks, you also need to add a Maven dependency to the 
`configuration-as-code` plugin, as shown above.

**Running benchmarks from Maven**

To easily run benchmrks from Maven, a Maven profile to run the benchmarks has been created
and is available starting Plugin-POM version 3.45. You can then run your benchmarks from the
command line using `mvn test -Dbenchmark`.

**Running benchmarks on ci.jenkins.io**

If you have your plugins hosted on ci.jenkins.io, you can easily run benchmarks directly from your Jenkinsfile

> by using the `runBenchmarks()` method after the `buildPlugin()` step in your which is now available in Jenkins

Reviewer: Add link to the Jenkins Pipeline Library?

Pipeline library. This function also accepts the path to your generated JMH benchmark reports as an optional 
parameter and archives the benchmark results. Running benchmarks in pull request builds allows you to constantly 
monitor the performance implications of a given change. For example, the Jenkinsfile from Role Strategy Plugin:
[source, groovy]
----
buildPlugin()
runBenchmarks('jmh-report.json')
----

> **Visualizing benchmark results**

Reviewer: Maybe? 

```suggestion
== Visualizing benchmark results
```

Benchmark reports generated (in JSON) can be visualized using the either the plugin:jmh-report[JMH Report Plugin]

> or by passing the JSON benchmark Reports to link:https://jmh.morethan.io[JMH visualizer]. As an example, here is

Reviewer: ```suggestion
or by passing the JSON benchmark Reports to link:https://jmh.morethan.io[JMH visualizer] web service. As an example, here is
```

a report of some benchmarks from the Role Strategy Plugin visulized using JMH visualizer.

> image::/images/post-images/jmh-microbenchmark-framework/jmh-visualizer.png[Role Strategy Plugin benchmarks visualized by JMH Visualizer]

Reviewer: Maybe crop it to 2 or 3 plots? Just to save size

> to the plugin and shows how even seemingly small changes can have major performance implications.

Reviewer: ```suggestion
to the plugin and shows how even seemingly small changes can bring major performance improvements.
Micro-benchmarks help to find such hotspots and to estimate the impact of changes.
```

naming conventions, the benchmarks will not interfere with your JUnit tests.
* Benchmark methods need to be annotated by `@Benchmark` for JMH to detect them.
* Classes containing benchmarks are found automatically by the `BenchmarkFinder`
when annotated with `@JmhBenchmark`.
* A reference to the Jenkins instance is available through either `JmhBenchmarkState#getJenkins()` or through
`Jenkins.getInstance()` like you would otherwise do. 
* `JmhBenchmarkState` provides `setup()` and `tearDown()` methods which can be overridden to configure the
 Jenkins instance according to your benchmark's requirements.
* The benchmark builds on ci.jenkins.io are currently throttled because of the limited availability of `highmem` nodes.

== Links and Feedback

If you have any feedback, comments or questions, please feel free to reach out to me through either
the link:https://gitter.im/jenkinsci/role-strategy-plugin[Role Strategy Plugin Gitter chat] or through 
the link:mailto:jenkinsci-dev@googlegroups.com[Jenkins Developer Mailing list].

* link:https://docs.google.com/presentation/d/e/2PACX-1vQXca_ZQNwI4vQ25Nw7lMSnSh4WBwKbC9VltT-7tOjS8cE69zMb2bgEbhgwurb1xA[Presentation slides]

> * link:https://www.youtube.com/watch?v=lyfbmhQd0Ag[Demo in Platform SIG meeting]

Reviewer: Maybe makes sense to add hyperlink to a second when your talk starts

* Documentation for the micro-benchmark framework:

> ** link:https://github.com/jenkinsci/jenkins-test-harness/blob/master/docs/jmh-benchmarks.adoc[Jenkins Test Harness]

Reviewer: ```suggestion
** Writing benchmarks: link:https://github.com/jenkinsci/jenkins-test-harness/blob/master/docs/jmh-benchmarks.adoc[Jenkins Test Harness]
```

> ** link:https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/benchmarks/jmh-benchmarks.md[Configuration-as-Code Plugin]

Reviewer: ```suggestion
** link:https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/benchmarks/jmh-benchmarks.md[Preconfiguring benchmarks with the JCasC Plugin]
```

> ** link:https://github.com/jenkinsci/plugin-pom#running-benchmarks[Plugin POM]

Reviewer: ```suggestion
** link:https://github.com/jenkinsci/plugin-pom#running-benchmarks[Running benchmarks with Plugin POM]
```
